src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,465 @@
+        } else if (mode == Mode.MODULE_CULLED) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
+            daiDebt = vat.gem(ilk, address(pool));
+        } else {
+            // MCD caged
+            // debt is obtained from free collateral owned by the End module
+            end.skim(ilk, address(pool));
+            daiDebt = vat.gem(ilk, address(end));
+        }
+
+        // Unwind amount is limited by how much:
+        // - max reduction desired
+        // - liquidity available
+        // - gem we have to withdraw
+        // - dai debt tracked in vat (CDP or free)
+        uint256 amount = _min(
+                            _min(
+                                _min(
+                                    supplyReduction,
+                                    availableLiquidity
+                                ),
+                                assetBalance
+                            ),
+                            daiDebt
+                        );
+
+        // Determine the amount of fees to bring back
+        uint256 fees = 0;
+        if (assetBalance > daiDebt) {
+            fees = assetBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
+            }
+        }
+
+        if (amount == 0 && fees == 0) {
+            emit Unwind(ilk, 0);
+            return;
+        }
+
+        require(amount <= 2 ** 255, "DssDirectDepositHub/overflow");
+
+        // To save gas you can bring the fees back with the unwind
+        uint256 total = _add(amount, fees);
+        pool.withdraw(total);
+        daiJoin.join(address(pool), total);
+
+        // normalized debt == erc20 DAI to pool (Vat rate for this ilk fixed to 1 RAY)
+
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(pool), address(pool), address(pool), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(pool), vow, _mul(fees, RAY));
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(pool), vow, _mul(total, RAY));
+        } else {
+            // This can be done with the assumption that the price of 1 aDai equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, address(end), -int256(amount));
+            vat.move(address(pool), vow, _mul(total, RAY));
+        }
+
+        emit Unwind(ilk, amount);
+    }
+
+    function exec(bytes32 ilk_) external {
+        Ilk memory ilk = ilks[ilk_];
```

Comment:
Doing this it's making 4 SLOADs when it is not necessary. Just one of the variables stored is used in all the paths and one extra in one particular path.